components/actions/ActionHero.tsx

```diff
@@ -150,65 +151,98 @@ const ActionNumber = styled.span`
+};
+
+function Crumb({ category }: { category: PartialCategory }) {
+  const id = `crumb-${category.id}`;
+  const ariaId = `tt-content-${category.id}`;
+  const isTruncated = category.name.length > MAX_CRUMB_LENGTH;
+  const name = isTruncated
+    ? `${category.name.slice(0, MAX_CRUMB_LENGTH).trim()}...`
+    : category.name;
+
+  return (
+    <>
+      <span id={id} aria-describedby={isTruncated ? ariaId : undefined}>
+        {category.url ? (
+          <Link href={category.url} passHref>
+            {name}
+          </Link>
+        ) : (
+          name
+        )}
+        {` / `}
+      </span>
+
+      {isTruncated && (
+        <UncontrolledTooltip
+          target={id}
+          id={ariaId}
+          placement="top"
+          role="tooltip"
+          trigger="focus hover"
+        >
+          {category.name}
+        </UncontrolledTooltip>
+      )}
+    </>
+  );
+}
+
+function ActionCategories({ categories }: { categories: Category[] }) {
   const plan = usePlan();
   const showIdentifiers =
     !plan.primaryActionClassification?.hideCategoryIdentifiers;
-  const displayCategories = [];
   const primaryCT = plan.primaryActionClassification;
   const primaryCatId = primaryCT?.id;
 
-  categories.forEach((cat, indx) => {
-    if (cat.type.id !== primaryCatId) return;
-    displayCategories[indx] = {};
-    let categoryTitle = cat.name;
-    if (cat.categoryPage) {
-      displayCategories[indx].url = cat.categoryPage.urlPath;
-      if (cat.identifier && showIdentifiers)
-        categoryTitle = `${cat.identifier}. ${cat.name}`;
-    } else if (primaryCT) {
-      displayCategories[
-        indx
-      ].url = `/actions?cat-${primaryCT.identifier}=${cat.id}`;
-    }
-    displayCategories[indx].name = categoryTitle;
-    displayCategories[indx].id = cat.id;
-    if (cat.parent) {
-      displayCategories[indx].parent = {};
-      let categoryParentTitle = cat.parent.name;
-      if (cat.parent.categoryPage) {
-        displayCategories[indx].parent.url = cat.parent.categoryPage.urlPath;
-        if (cat.parent.identifier && showIdentifiers) {
-          categoryParentTitle = `${cat.parent.identifier}. ${cat.parent.name}`;
-        }
-      } else {
-        displayCategories[
-          indx
-        ].parent.url = `/actions?cat-${primaryCT.identifier}=${cat.parent.id}`;
-      }
-      displayCategories[indx].parent.name = categoryParentTitle;
-      displayCategories[indx].parent.id = cat.parent.id;
-    }
-    return true;
-  });
+  const displayCategories: PartialCategory[] = categories
+    .filter((cat) => cat.type.id === primaryCatId)
+    .map((cat) => ({
+      id: cat.id,
+      name: getCategoryName(cat, showIdentifiers),
+      url: getCategoryUrl(cat, primaryCT),
+      parent: cat.parent
+        ? {
+            id: cat.parent.id,
+            name: getCategoryName(cat.parent, showIdentifiers),
+            url: getCategoryUrl(cat.parent, primaryCT),
+          }
+        : undefined,
+    }));
```

Comment:
@terotik This is what I mentioned in the daily, I'm a big fan of using `map` and small helper functions with a clear input and output for these kinds of trickier data mappings.